### PR TITLE
Infiniscroll for choices

### LINKED ISSUE #15: Page size and infiniscroll in single choice and multichoice prompts
Large number of alternatives should not be displayed at once.

We should only display a configured N alternatives (call it page size, whatever), and allow infiniscroll for the rest. Same as what prompts.js does.

The options should be configurable and passed from Prompt constructor.

CleanShot_2024-11-25_at_14 09 33

### PR DESCRIPTION
**Pull request by lbialy**
closes #15

### CODE REVIEW

File: modules/core/src/main/scala/Prompt.scala [hunk 1]
```diff
@@ -52,9 +62,16 @@ object Prompt:
     override def handler(
         terminal: Terminal,
         output: Output,
-        colors: Boolean
+        colors: Boolean,
+        windowSize: Int = 10
```
Comment (keynmol) [inline modules/core/src/main/scala/Prompt.scala:66]: Is there a reason to not push windowSize parameter into just the prompts that need it? namely single choice and multi-choice.

Unlike colors, it's not a universal property shared by all prompts, it's only really applicable to single/multi choice and (future) search prompt

File: modules/cats-effect/src/main/scala/PromptsIO.scala [hunk 2]
```diff
@@ -22,15 +22,16 @@ import cue4s.*
 class PromptsIO private (
     protected val out: Output,
     protected val terminal: Terminal,
-    protected val colors: Boolean
+    protected val colors: Boolean,
+    protected val windowSize: Int
```
Comment (keynmol) [inline modules/cats-effect/src/main/scala/PromptsIO.scala:26]: Is there a reason to not push windowSize parameter into just the prompts that need it? namely single choice and multi-choice. 

Unlike colors, it's not a universal property shared by all prompts, it's only really applicable to single/multi choice and (future) search prompt

Comment (lbialy) [reply modules/cats-effect/src/main/scala/PromptsIO.scala:26]: wouldn't it make sense to treat it as the universal maximum viewport height? maybe it also makes sense for text input (which does not implement scrolling atm)

Comment (keynmol) [reply modules/cats-effect/src/main/scala/PromptsIO.scala:26]: At this point frankly I don't see a reason for windowSize on input – we don't have the logic for trimming viewport on text input, nor do we have currently any precedent showing it's necessary.

Let's keep it only as part of the prompts for now, and later if necessary we can group them into `Options` and make it more of a first class citizen.

Comment (lbialy) [reply modules/cats-effect/src/main/scala/PromptsIO.scala:26]: ok

Comment (keynmol) [reply modules/cats-effect/src/main/scala/PromptsIO.scala:26]: I think you removed the parameter usages, but kept it in the constructor

File: modules/cats-effect/src/main/scala/PromptsIO.scala [hunk 3]
```diff
@@ -48,7 +49,9 @@ object PromptsIO:
   def apply(
       out: Output = Output.Std,
       createTerminal: Output => Terminal = Terminal.ansi,
-      colors: Boolean = true
+      colors: Boolean = true,
+      windowSize: Int = 10
   ): Resource[IO, PromptsIO] = Resource.fromAutoCloseable(
-    IO(new PromptsIO(out, createTerminal(out), colors))
+    IO(new PromptsIO(out, createTerminal(out), colors, windowSize))
```
Comment (keynmol) [inline modules/cats-effect/src/main/scala/PromptsIO.scala:55]: Same here

File: modules/cats-effect/src/main/scala/PromptsIO.scala (hunk 2 above)
Comment (lbialy) [reply modules/cats-effect/src/main/scala/PromptsIO.scala]: fixed

File: modules/cats-effect/src/main/scala/PromptsIO.scala (hunk 3 above)
Comment (lbialy) [reply modules/cats-effect/src/main/scala/PromptsIO.scala]: fixed

File: modules/core/src/test/scala/ExampleTests.scala [hunk 4]
```diff
@@ -32,6 +32,30 @@ class ExampleTests extends munit.FunSuite, TerminalTests:
...
 
+  terminalTestComplete("alternatives.infiniscroll.single")(
+    Prompt.SingleChoice(
+      "How was your day?",
+      List(
+        "amazing",
+        "productive",
+        "relaxing",
+        "stressful",
+        "exhausting"
```
Comment (keynmol) [inline modules/core/src/test/scala/ExampleTests.scala:43]: Can you please add a test which also has some typing? To make sure that filtering doesn't screw up the pagination

**Review by keynmol — APPROVED** [review]
This looks great!

Could use some more tests, but we can add it later, I created a separate issue for it.

**Pull request merged**